ex3/m_depontes/main.py

```diff
@@ -0,0 +1,473 @@
+        w = np.zeros(n)  # 初期重みをゼロで初期化
+
+        for iteration in range(max_iter):
+            w_old = w.copy()
+            for j in range(n):
+                # 残差を計算
+                residual = y - (X @ w - X[:, j] * w[j])
+                # 重みの更新
+                rho = X[:, j].T @ residual
+                l1_term = alpha * l1_ratio
+                l2_term = alpha * (1 - l1_ratio)
+                if rho < -l1_term:
+                    w[j] = (rho + l1_term) / (X[:, j].T @ X[:, j] + l2_term)
+                elif rho > l1_term:
+                    w[j] = (rho - l1_term) / (X[:, j].T @ X[:, j] + l2_term)
+                else:
+                    w[j] = 0
+
+            # 収束判定
+            if np.linalg.norm(w - w_old, ord=2) < tol:
+                break
+
+        return w
+
+    def fit(self) -> None:
+        """データを読み込み、線形回帰を行う関数.
+
+        入力：
+            path(str): データファイルのパス
+            M(int): 次数
+
+        出力：
+            w(np.ndarray): 重み, shape = (次数)
+        """
+        self.data = read_csv(self.path)
+
+        self.w = self.OLS(self.M, self.data)
+
+    def predict(self) -> np.ndarray:
+        """予測を行う関数.
+
+        入力：
+            data(np.ndarray): 入力値, shape = (入力値,入力値に対する出力値)
+            M(int): 次元数
+
+        出力：
+            y_pred(np.ndarray): 予測値, shape = (次数)
+            data(np.ndarray): 入力値, shape = (入力値,入力値に対する出力値)
+        """
+        x = self.data[0]
+        X = np.vstack([x**i for i in range(self.w.shape[0])]).T
+        y_pred = (X @ self.w).flatten()  # 予測値
+
+        return self.data, y_pred
+
+
+class PCA:
+    """主成分分析の実装を行うクラス.
+
+    メンバ関数：
+        compute_S: 観測値ベクトルの集合およびその期待値ベクトルから変動行列を計算する関数
+        compute_sorted_eigen: 分散共分散行列から固有値および固有ベクトルを計算し，それぞれ固有値の降順に並べて返す関数
+        fit: 分散最大基準に基づき，変換行列を学習する関数
+        transform: 学習した変換行列によって特徴量ベクトルを射影する関数
+        explained_variance_ratio: 累積寄与率を計算する関数
+    属性：
+        n_components(int): 主成分の数
+        A(np.ndarray): 変換行列, shape = (次元数, 次元数)
+        eigenvalues(np.ndarray): 固有値, shape = (次元数,)
+        eigenvectors(np.ndarray): 固有ベクトル, shape = (次元数, 次元数)
+        means(np.ndarray): 期待値ベクトル, shape = (次元数)
+    """
+
+    def __init__(
+        self,
+        n_components: int = 2,
+    ):
+        """初期化関数.
+
+        入力：
+            n_components(int): 主成分の数
+        出力：
+            None
+        """
+        self.n_components = n_components
+        self.A = None
+        self.eigenvalues = None
+        self.eigenvectors = None
+        self.means = None
+
+    def compute_S(
+        self,
+        X: np.ndarray,
+        means: np.ndarray,
+    ) -> np.ndarray:
+        """観測値ベクトルの集合およびその期待値ベクトルから変動行列を計算する関数.
+
+        入力：
+            X(np.ndarray): 観測値ベクトルの集合, shape = (観測値ベクトルの数, 次元数)
+            means(np.ndarray): 期待値ベクトル, shape = (次元数)
+
+        出力：
+            S(np.ndarray): 変動行列, shape = (次元数, 次元数)
+        """
+        S = 0
+        for i in range(len(X)):
+            S += np.outer((X[i] - means), (X[i] - means).T)
```

Comment:
```suggestion
        S = (X - means) @ (X - means).T
```
ここは行列計算で1行で書くことができます。